design-documents/bit-precise-types.rst

```diff
@@ -0,0 +1,506 @@
+
+There are two decisions around the representation of a "small" ``_BitInt`` that
+we have identified.  (1) Whether required bits are stored in the least
+significant end or most significant end of a register or region in memory. (2)
+Whether the "remaining" bits after rounding up to the size specified in
+`Alignment and sizes`_ are specified or not.  The choice of *how* "remaining"
+bits would be specified would tie in to the choice made for (1).
+
+
+Options and their trade-offs
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+We have identified three viable options:
+
+A. Required bits stored in most significant end.
+   Not-required bits are specified as zero at ABI boundaries.
+B. Required bits stored in least significant end.
+   Not-required bits are unspecified at ABI boundaries.
+C. Required bits stored in least significant end.
+   Not-required bits are specified as zero- or sign-extended.
+
+While it would be possible to make different requirements for bit-precise
+integer types in memory vs in registers, we believe that the combined negatives
+of the choice are reason enough to not look into the option.  These negatives
+being that code would have to perform a transformation on loading and storing
+values, and that different representations in memory and registers is likely to
+cause programmer confusion.
+
+Similarly, it would be possible to define a representation in registers that
+does something like specifying bits ``[2-7]`` of a ``_BitInt(2)`` but leaves
+bits ``[8-63]`` unspecified.  This would seem to choose the worst of both worlds
+in terms of performance, since one must both ensure "overflow" from an addition
+of ``_BitInt(2)`` types does not affect the specified bits **and** ensure that
+the unspecified bits above bit number 7 do not affect multiplication or division
+operations.  Hence we do not look at variations of this kind.
+
+For option ``A`` there is an extra choice around how "large" values are stored.
+One could either have the "padding" bits in the least significant "chunk", or
+the most significant "chunk".  Having these padding bits in the least
+significant chunk would mean require something like a widening cast would
+require updating every "chunk" in memory, hence we assume large values of option
+``A`` would be represented with the padding bits in the most significant chunk.
+
+
+Option ``A`` has the following benefits:
+
+- For small values in memory, on AArch64, the operations like ``LDADD`` and
+  ``LD{S,U}MAX`` both work (assuming the relevant register operand is
+  appropriately shifted).
+
+- Operations ``+,-,%,==,<=,>=,<,>,<<`` all work without any extra instructions
+  (which is more of the common operations than other representations).
+
+It has the following negatives:
+
+- This would be a less familiar representation to programmers.  Especially the
+  fact that a ``_BitInt(8)`` would not have the same representation in a
+  register as a ``char`` could cause confusion (e.g. when debugging, or writing
```

Comment:
```suggestion
  register as a ``char`` could cause confusion (For example, when debugging, or writing
```

Reply:
For some reason the "for example" reads clunkier than "e.g." to me.
I went with `cause confusion (we imagine when debugging or writing assembly code)`.
Hopefully that avoids the ease-of-parsing difficulty for non-native speakers.